src/fstar/FStar.CheckedFiles.fs

```diff
@@ -0,0 +1,335 @@
+                    fn module_name)
+    in
+    match digest with
+    | Inl msg -> Inl msg
+    | Inr dig ->
+      hash_deps ((Dep.lowercase_module_name fn, dig) :: out) deps
+  in
+  hash_deps [] binary_deps
+
+(*
+ * Load a checked file into mcache
+ *
+ * This is loading the parsing data, and tc data as Unknown (unless checked file is invalid)
+ *
+ * See above for the two steps of loading the checked files
+ *)
+let load_checked_file (fn:string) (checked_fn:string) :unit =
+  if checked_fn |> BU.smap_try_find mcache |> is_some then ()  //already loaded
+  else
+    if not (BU.file_exists checked_fn)
+    then let msg = BU.format1 "checked file %s does not exist" checked_fn in
+         BU.smap_add mcache checked_fn (Invalid msg, Inl msg)
+    else match BU.load_value_from_file checked_fn with
+         | None ->
+           let msg = BU.format1 "checked file %s is corrupt" checked_fn in
+           BU.smap_add mcache checked_fn (Invalid msg, Inl msg)
+         | Some (vnum, deps_dig, dig, parsing_data, tc_result) ->
+           if vnum <> cache_version_number
+           then let msg = BU.format1 "checked file %s has incorrect version" checked_fn in
+                BU.smap_add mcache checked_fn (Invalid msg, Inl msg)
+           else let current_dig = BU.digest_of_file fn in
+                if dig <> current_dig
+                then begin
+                  if Options.debug_any () then
+                  BU.print4 "Checked file %s is stale since incorrect digest of %s, \
+                    expected: %s, found: %s\n"
+                    checked_fn fn current_dig dig;
+                  let msg = BU.format2 "checked file %s is stale (digest mismatch for %s)" checked_fn fn in
+                  BU.smap_add mcache checked_fn (Invalid msg, Inl msg)
+                end
+                else BU.smap_add mcache checked_fn (Unknown (deps_dig, dig, tc_result), Inr parsing_data)
+
+(*
+ * Second step for loading checked files, validates the tc data
+ *)
+let load_checked_file_with_tc_result (deps:Dep.deps) (fn:string) (checked_fn:string) :unit =
+  load_checked_file fn checked_fn;  //first step, in case some client calls it directly
+  match BU.smap_try_find mcache checked_fn with
+  | None -> failwith "Impossible, load_checked_file must add an entry to mcache"
+  | Some (Invalid _, _)
+  | Some (Valid _, _) -> ()  //already marked Valid or Invalid
+  | Some (Unknown (deps_dig, dig, tc_result), parsing_data) ->
+    match hash_dependences deps fn with
+    | Inl msg ->
+      BU.smap_add mcache checked_fn (Invalid msg, parsing_data)
+    | Inr deps_dig' ->
+      if deps_dig = deps_dig'
+      then begin
+        //mark the tc data of the file as valid
+        BU.smap_add mcache checked_fn (Valid (dig, tc_result), parsing_data);
+        //if there exists an interface for it, mark that too as valid
+        //validity of implementaton tc data implies validity of iface tc data
```

Comment:
Is it really worth doing this? It seems to duplicate the logic of the Unknown, Valid state machine for interface files. Since all of this is cached anyway, couldn't we just remove this code and validate an interface when it is explicitly requested?